### PR TITLE
Implement R.collectBy

### PR DESCRIPTION
**Pull request by customcommander**
Follow up to #2706 

I wanted to keep the implementation to a bare (perhaps naive) minimum and get some early feedback. I'm not too sure about a few things:

1. Compatibility with transducers.
2. Compatibility with any relevant fantasy land specs.
3. I'm not too familiar with the Hindley-Milner notation.
4. Is it something that the collective would consider a nice addition to the library? :) 

Thank you

### CODE REVIEW

File: source/collectBy.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import values from './values';
+
+/**
+ * Splits a list into sub-lists, based on the result of calling a String-returning function on each element,
+ * and grouping the results according to values returned.
+ *
+ * @func
+ * @memberOf R
+ * @category List
+ * @sig (a -> String) -> [a] -> {String: [a]}
```
Comment (customcommander) [inline source/collectBy.js]: I'm not familiar with the Hindley-Milner notation

File: source/collectBy.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * @param {Function} fn Function :: a -> String
+ * @param {Array} list The array to split into sub-arrays.
+ * @return {Array}
+ *    An array of arrays where each sub-array contains items for which
+ *    the String-returning function has returned the same value.
+ * @see R.groupBy
```
Comment (customcommander) [inline source/collectBy.js]: Should `R.partition` be mentioned?

Comment (CrossEye) [reply source/collectBy.js]: Yes, I think that's a good idea.  It's quite related.  In fact, `collectBy(pred)` will group a list into the same groups that `partition(pred)` does, although the order might not be the same.

File: source/collectBy.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ * @example
+ *      R.collectBy(R.prop('type'), [
+ *        {type: 'breakfast', item: '☕️'},
+ *        {type: 'lunch', item: '🌯'},
+ *        {type: 'dinner', item: '🍝'},
+ *        {type: 'breakfast', item: '🥐'},
+ *        {type: 'lunch', item: '🍕'}
+ *      ]);
```
Comment (CrossEye) [inline source/collectBy.js]: I think these should be in a slightly different order so that the result could not be mistaken for something equivalent to `compose(transpose, splitEvery(3))`

File: test/collectBy.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+      [{type: 'breakfast', item: '☕️'}, {type: 'breakfast', item: '🥐'}],
+      [{type: 'lunch', item: '🌯'}, {type: 'lunch', item: '🍕'}],
+      [{type: 'dinner', item: '🍝'}]
+    ]);
+  });
+
+  it('returns an empty list if given an empty array', function() {
+    eq(R.collectBy(R.prop('x'), []), []);
+  });
+
```
Comment (CrossEye) [inline test/collectBy.js]: I think we might want a test-case with a function that returns numbers.  (Maybe my modulus example above.)

File: source/collectBy.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ *      // [ [ {type: 'breakfast', item: '☕️'},
+ *      //     {type: 'breakfast', item: '🥐'} ],
+ *      //   [ {type: 'lunch', item: '🌯'},
+ *      //     {type: 'lunch', item: '🍕'} ],
+ *      //   [ {type: 'dinner', item: '🍝'} ] ]
+ */
+var collectBy = _curry2(function collectBy(fn, list) {
+  return values(groupBy(fn, list));
```
Comment (CrossEye) [inline source/collectBy.js]: If you have some time, could you compare the performance to a iteration-based approach?  There may not be much difference, in which case, we can keep this simple code.  But if the difference is substantial (perhaps: a factor of more than 2), we should probably switch to more performant code.

Comment (customcommander) [reply source/collectBy.js]: How do you usually do this? Do you use any benchmarking library?

File: test/collectBy.js (hunk 4 above)
Comment (customcommander) [reply test/collectBy.js]: Makes sense

File: source/collectBy.js (hunk 3 above)
Comment (customcommander) [reply source/collectBy.js]: Ok will do

File: source/collectBy.js (hunk 1 above)
Comment (customcommander) [reply source/collectBy.js]: sig tag updated with your suggestion

File: source/collectBy.js (hunk 2 above)
Comment (customcommander) [reply source/collectBy.js]: Mentioned `partition` in see tag

File: source/collectBy.js (hunk 3 above)
Comment (customcommander) [reply source/collectBy.js]: Added more objects and "shuffled" the array to avoid seeing patterns where there are none. What do you think?

File: source/collectBy.js (hunk 5 above)
Comment (customcommander) [reply source/collectBy.js]: If a grouping function returns a non-string value must cause `collectBy` to return an empty list, then I'd probably have to implement with classic loop statements. Do you still want to me to measure the performance in that case and what would be an acceptable threshold?

File: test/collectBy.js (hunk 4 above)
Comment (CrossEye) [reply test/collectBy.js]: I am starting to find review conversations harder to follow on GitHub.  I thought I had made a suggestion for a simpler example, but I can't find it now.  The version of this in the test suite seems perfect, but here, something like this might be simpler:

```js
collectBy (n => n % 10, [5, 11, 12, 22, 23, 25, 32, 35, 43])
//=> [[11], [12, 22, 32], [23, 43], [5, 25, 35]]
```

File: source/collectBy.js (hunk 5 above)
Comment (CrossEye) [reply source/collectBy.js]: > Do you use any benchmarking library?

Usually a quick test in [jsperf][jp] or [JSBench.me][jb] is plenty.  This doesn't have to live long.

  [jb]:https://jsbench.me/
  [jp]: https://jsperf.com/

Comment (CrossEye) [reply source/collectBy.js]: I would simply use the default object behavior for keys, which should take no effort: 

```js
const foo = {}
foo[true] = 1
foo[{a: 1}] = 2
foo[/[a-z]/] = 3
foo[(n) => n * n] = 4
// etc.

foo //=>
// {
//   "/[a-z]/": 3,
//   "[object Object]": 2,
//   "function (n) {\n  return n * n;\n}": 4, // (or whatever)
//   "true": 1
// }
```

If you peruse our source code, many functions which could easily be written in terms of other ones are written natively, often with `while` loops.  The rationale is that the library does this nasty stuff so you don't have to.

**Comment by customcommander** [comment]
@CrossEye I'll finish this!

**Comment by CrossEye** [comment]
@customcommander:

I'd forgotten all about it, I'm afraid.

File: test/collectBy.js (hunk 4 above)
Comment (customcommander) [reply test/collectBy.js:24]: I resorted to property-based testing since you `jsverify` is available.

File: source/collectBy.js (hunk 5 above)
Comment (customcommander) [reply source/collectBy.js]: I ended up writing the function "natively". Would you still see some performance benchmark?

**Comment by customcommander** [comment]
@CrossEye There you go :-) I resolved all comments. Can you check and see if there's room for improvements?

File: source/collectBy.js (hunk 5 above)
Comment (CrossEye) [reply source/collectBy.js]: No.  The only reason for that was that if we were skipping our usual style, I wanted to ensure that it didn't hurt performance.

File: test/collectBy.js (hunk 4 above)
Comment (CrossEye) [reply test/collectBy.js:24]: That's fine.  I have very mixed feelings about property-based testing.  It's clearly a better way to catch edge cases, and that's the most important facet of unit tests.  But a strong secondary feature of unit tests is to serve as an extra API guide.  When I look to see how a project or a new feature works, I usually look first to the test suite rather than the docs.  That doesn't work nearly as well with property-based testing.

**Comment by CrossEye** [comment]
:herb:

This looks great to me.  Anyone else? @ramda/core

File: source/collectBy.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import _reduce from './internal/_reduce';
+
+/**
+ * Splits a list into sub-lists, based on the result of calling a String-returning function on each element,
+ * and grouping the results according to values returned.
+ *
+ * @func
+ * @memberOf R
+ * @category List
+ * @sig (a -> String) -> [a] -> [[a]]
```
Comment (hitmands) [inline source/collectBy.js:11]: >  based on the result of calling a String-returning function on each element

I am not sure about the constraints of a string returning function, this might be required for `groupBy` but seems to not be strictly needed here.

`collectBy` really looks like `partition` without the need for the returned tuple to be a pair.

```ts
// Tuple length is 2
// eq to partition as fn is a predicate
const [evens, odds] = R.collectBy((n) => n % 2 === 0, [1, 2, 3, 4, 5]);
``` 

```ts
const bySpeed = (speed) => {
  if (speed < 30) { return 0; }
  if (speed < 90) { return 1; }
  return 2;
};

const [slow, average, fast] = R.collectBy(bySpeed, [10, 30, 70, 250]);
```

Probably better to reword

File: source/collectBy.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    var tag = fn(x);
+    if (o[tag] === undefined) { o[tag] = []; }
+    o[tag].push(x);
+    return o;
+  }, {}, list);
+  var newList = [];
+  for (var tag in group) { newList.push(group[tag]); }
+  return newList;
+});
+export default collectBy;
```
Comment (hitmands) [inline source/collectBy.js:44]: this is more of a question, 
but why the implementation isn't aliasing `groupBy . values` ?

File: source/collectBy.js (hunk 6 above)
Comment (customcommander) [reply source/collectBy.js:11]: I _think_ the string requirement is just a clue for the developer. Related items are grouped under the same key of an object and keys are always strings. The function can indeed return anything but it will be automatically casted into a string.

@CrossEye and I did discuss this already. See https://github.com/ramda/ramda/pull/2715#discussion_r236656183

File: source/collectBy.js (hunk 7 above)
Comment (customcommander) [reply source/collectBy.js:44]: This was actually the case in my first attempt. I believe @CrossEye had some questions about the performance and hinted at implementing from "scratch".

File: source/collectBy.js (hunk 6 above)
Comment (hitmands) [reply source/collectBy.js:11]: sounds fair, in the end it's more about wording.

I'd personally prefer "A function returning a valid property key".

- [Typescript / PropertyKey](https://github.com/microsoft/TypeScript/blob/master/src/lib/es5.d.ts#L85)
- [Symbols as PropertyKey](https://2ality.com/2014/12/es6-symbols.html#symbols-as-property-keys)

Comment (CrossEye) [reply source/collectBy.js:11]: I missed this thread when I merged.  @hitmands: I think you're right about the wording.  We should probably change it that way.  Somewhere else we use a signature like

```hs
somefunc :: Index a => (b -> a) -> ?? -> Object
Index == String | Int | Symbol
```

We should probably get that cleaned up and make this a habit for describing functions that return Object keys.  And the description should match.

That can be a separate task, though.
